{% extends 'base.html' %}


{% block content %}

<div class="container">

  
<h1 class="bst"> {% block title %} Changelog {% endblock %} </h1>


<div class="underline">v 1.8.2 (26.02.2024)</div>
<div class="btn-pvs">!! The automatic classification algorithm is currently under construction. HerediVar is already prepared for it. Thus, you might encounter disabled buttons or dummy data. It will be enabled in a future update</div>
<div class="btn-pvs">!! Scores from likelihood ratio tests are currently under construction. A simple reannotation is required to show them on HerediVar once they are ready to be downloaded from HerediCaRe.</div>
<div>
    Bugfixes:
    <ul>
        <li>Fixed a bug where the SpliceAI score was not shown correctly when it is "."</li>
        <li>In rare cases the submission condition in ClinVar submissions is na with a description. This case is now handled properly and does not produce erroneous links.</li>
        <li>Fixed a bug when the assay comment for splicing assays could not be inserted due to length</li>
    </ul>
</div>

<div class="underline">v 1.8.1 (23.02.2024)</div>
<div>
    General changes:
    <ul>
        <li>Changed is_selected for classifications to a more descriptive state information in downloaded vcf files</li>
    </ul>
    Bugfixes:
    <ul>
        <li>Fixed a bug when there were no HerediCaRe center classifications but a consensus classification from HerediCaRe</li>
        <li>A selected criterium would automatically uncheck if a mutually exclusive criterium was unselected and the mutually exclusive criterium itself is mutually exclusive with the selected criterium</li>
        <li>Fixed default strength for supporting criteria in ACMG standard scheme</li>
        <li>Fixed wrapping of PVS1_mod criterium</li>
        <li>Fixed a bug where the SpliceAI scores would not show on longer indels</li>
    </ul>
</div>
<div class="underline">v 1.8 (21.02.2024)</div>
<div>
    General changes:
    <ul>
        <li>Added scores from likelihood ratio tests. </li>
        <li>Removed BayesDel annotations from TP53 DB, because they are annotated from a dedicated BayesDel anntation anyway.</li>
        <li>Improved Cancerhotspots internal representation, display & representation in the INFO column of vcfs</li>
        <li>Added HerediCaRe center specific classifications</li>
        <li>Added CSpec BRCA1/2 assays to annotation algorithm</li>
        <li>Copy evidence of user classifications to a consensus classification now also automatically selects the state correctly</li>
    </ul>
    Bugfixes:
    <ul>
        <li>Fixed a bug where some variants got raw SpliceAI scores instead of the masked ones</li>
        <li>When constructing an URL for the variant details page it was previously not possible to omit the "chr" prefix for the chromosome</li>
        <li>Task force protein domains are now shown correctly</li>
        <li>Transcripts on the negative strand were shown as if they were on the positive strand in the IGV browser</li>
        <li>Fixed a bug when the user clicked on logout even though the user was already logged out (for example because of inactivity)</li>
        <li>The default scheme was not properly working. It would automatically select the last scheme in the list as default no matter what was selected on the admin dashboard</li>
        <li>Fixed a bug when updating ClinVar submissions</li>
        <li>When selecting any BP criterium the enigma BRCA1/2 schemes would instantly switch to class 2. This is now fixed and only BP1_str triggers class 2</li>
        <li>Previously the class by scheme was not updated correctly when the user updated their classification</li>
        <li>Fixed a bug in the criteria column of the automatic classification results that would shrink to width zero when the screen is resized</li>
        <li>Fixed a bug where unselected criteria where shown as selected for consensus classifications</li>
    </ul>
</div>

<div class="bst bsb">
    <h4>Changelog</h4>
    <div class="underline">v 1.7 (16.01.2024)</div>
    <div>
        General changes:
        <ul>
            <li>Added form to create structural variants: DEL, DUP, INV</li>
            <li>Added structural variants to browse variants table</li>
            <li>Added START, END, SVTYPE, IMPRECISE info columns to vcf with structural variants</li>
            <li>Added the general comment to the classified variants download</li>
            <li>Added a separate download link for all classified structural variants</li>
            <li>Added structural variants track to IGV</li>
            <li>Added the coldspot annotation</li>
            <li>Added variant search which allows the user to search directly for a specific variant using this format: chrom-pos-ref-alt OR chrom-start-end-sv_type in case of structural variants</li>
            <li>Changed External ID search to overall ID search. It now also allows to search for heredivar variant ids.</li>
            <li>Update search help</li>
            <li>Added variant type search option</li>
            <li>Updated ClinVar to v. 20240207</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Fixed tooltips in browse variant table</li>
            <li>The annotation service would fail when the variant does not have a dbsnp rsid</li>
            <li>Fixed a bug that increased the size of the consensus classification label for class 4M</li>
            <li>The clinvar variation was in the downloaded vcf files twice. Removed one of the copies. The clinvar variation id info is now called "clinvar"</li>
            <li>In rare cases ClinVar submissions are missing which would result in an error when downloading the vcf file. In such cases the clinvar_submissions info in the downloaded vcf file will be set to None</li>
            <li>Fixed a bug that prevented the user to change the state of criteria when using microsoft edge</li>
        </ul>
        Known issues:
        <ul>
            <li>HerediCare Center specific classifications are not available</li>
            <li>HerediCare priors, log likelihood, segregation & cooccurance are not available</li>
        </ul>
    </div>
    <div class="underline">v 1.6.1 (11.01.2024)</div>
    <div>
        Bugfixes:
        <ul>
            <li>Searching for 4M criterium with include_heredicare option would result in an error</li>
            <li>Searching for class 4 consensus classification with include_heredicare option previously resulted in consensus class 4 and heredicare class 3+</li>
        </ul>
    </div>

    <div class="underline">v 1.6 (10.01.2024)</div>
    <div>
        General changes:
        <ul>
            <li>Classifications can now be submitted without selecting any criterium</li>
            <li>Added the ClinGen gene specific ACMG classification scheme for ATM</li>
            <li>Added page to hide classification schemes:
                <ul>
                    <li>disable schemes</li>
                    <li>set default scheme</li>
                </ul>
            </li>
            <li>Improved automatic selection of schemes based on genes</li>
            <li>Changed suffixes from criteria with differing strenghts</li>
            <li>Download of all classified variants now also contains all variants which were only consensus classified in heredicare. Also added a source of the classification info column</li>
            <li>Added gnomAD v4 link to variant banner</li>
            <li>Variant annotations can now be aborted from the admin dashboard</li>
            <li>Added a button to reannotate all aborted variants from the admin dashboard and moved the reannotate erroneous variants button to the (Re-)annotate tab of the admin dashboard</li>
            <li>Added automatic classification algorithm v1.0</li>
            <li>Added the automatic classification result to the browse variant table & variant details page</li>
            <li>Added automatic classification search</li>
            <li>Added a button to the admin dashboard to annotate all variants which are missing an annotation</li>
            <li>Updated VEP from 107.0 to version 110.1</li>
            <li>Added possibility to "unselect" a criterium when there is evidence that this criterium does not hold</li>
            <li>Added button to automatically select criteria based on the automatic classification algorithm</li>
            <li>Added automatic classification algorithm results to vcf export</li>
            <li>Added flag to vcf export of heredivar classifications if a criterium was selected or not selected in response of evidence</li>
            <li>Added all external ids to variant vcf export including for example heredicare VIDs, rsids, cosmic ids, etc.</li>
            <li>Improved display of intergenic variants on variant details page</li>
            <li>Changed wording in contact form & enabled contact form</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Fixed a bug where the criterium strength addition to the label would reset when the default strength was not its original strength.</li>
            <li>Fixed the order of selectable criteria strengths on the classify variants page</li>
            <li>Fixed a bug when submitting 4M classifications</li>
            <li>Previously, updates to ClinVar submissions were not working & removed the 'customAssertionScore' from ClinVar submissions</li>
            <li>Fixed a bug when trying to add all variants from a search to a list</li>
            <li>In BRCA2 v1.0.0 criteria supporting strength was not selectable</li>
            <li>Sometimes, the classification scheme was not changeable in consensus classifications</li>
        </ul>
    </div>

    <div class="underline">v 1.5 (17.11.2023)</div>
    <div>
        General changes:
        <ul>
            <li>Added option to include heredicare consensus classififcations in consensus classification search</li>
            <li>Added BayesDEL annotation</li>
            <li>Improved ClinVar review status & summary display</li>
            <li>Improved range search. Each range search entry must now be of the format: chrom&lt;sep&gt;start&lt;sep&gt;end. With &lt;sep&gt; being either :, -, space or tabulator. The chromosome number can start with 'chr', but you can also choose to omit it.</li>
            <li>Added external id search. Which allows to search for external ids like heredicare VIDs, COSMIC COSV ids or dbSNP RSIDs.</li>
            <li>Added cDNA search</li>
            <li>Update search help</li>
            <li>Added COSMIC COSV ids & moved COSMIC links from consequence table to the variant banner.</li>
            <li>Added GnomAD popmax allele count (AC) annotation</li>
            <li>Updated documentation</li>
            <li>Added annotation search</li>
            <li>The classification comment is now optional</li>
            <li>REVEL annotations are now transcript specific and show the maximal value in the overview. The details list REVEL scores for all annotations</li>
            <li>Additional warnings added to annotation service</li>
            <li>User classifications can now be deleted by the user that created it</li>
            <li>Renamed class 4 to 4M & changed its color to orange</li>
            <li>Renamed PALB2 v1.0.0 & TP53 v1.4.0 classification schemes</li>
            <li>Improved performance of transcript track in IGV</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Fixed a bug where the "gene", "your classification" and "consensus classification" columns were not clickable in the browse variant table</li>
            <li>Fixed a bug where in some cases some ClinVar submissions were not annotated</li>
            <li>Fixed a bug which caused some annotations to be erroneous, when importing variants from HerediCare</li>
            <li>Fixed status of deleted VIDs upon variant import</li>
            <li>In some cases the consequence was duplicated in the variant banner.</li>
            <li>Increased the size of the criteria display buttons, because the label did not fit into one line in some cases.</li>
            <li>Fixed a bug when submitting incomplete information on the "submit to ClinVar" page</li>
        </ul>
        Known issues:
        <ul>
            <li>HerediCare Center specific classifications are not available</li>
            <li>HerediCare priors, log likelihood, segregation & cooccurance are not available</li>
        </ul>
    </div>

    <div class="underline">v 1.4.1 (02.11.2023)</div>
    <div>
        Bugfixes:
        <ul>
            <li>Fixed a bug where the user was unable to insert classifications by the new ClinGen criteria</li>
        </ul>
    </div>
    <div class="underline">v 1.4 (27.10.2023)</div>
    <div>
        General changes:
        <ul>
            <li>Update landing page</li>
            <li>Moved previous changelog announcements to a new page</li>
            <li>The admin dashboard now allows reannotation of a user defined set of databases</li>
            <li>The admin dashboard now shows an overview of all annotations and annotation errors / warnings</li>
            <li>Added a contact form where unregistered people can ask a question about a variant</li>
            <li>Added suffix to criteria names if the selected strength is not the default strength</li>
            <li>Variants can now be hidden</li>
            <li>The variant list can now be sorted by chromosomal position and insertion date</li>
            <li>The page size in the variant list can now be adjusted</li>
            <li>Consensus classifications are now preselected from the most recent consensus classification for each scheme</li>
            <li>Added four new ClinGen ENIGMA schemes for specific genes: BRCA1 (v.1.0.0), BRCA2(v.1.0.0), TP53 (v.1.4.0), PALB2 (v.1.0.0)</li>
            <li>Changed SpliceAI annotations to masked ones</li>
            <li>Added variant import from HerediCare</li>
            <li>Added HerediCare number of families and heredicare number of cases annotations</li>
            <li>Improved criteria popover</li>
            <li>Added variant import history page</li>
            <li>Enabled HerediCare annotations in admin dashboard</li>
            <li>Added downloads page</li>
            <li>Added variant check page</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Variants are now not selected by default on the search variants page</li>
            <li>Abbreviations for criteria are now written with upper case letters</li>
            <li>Disabled criteria were not shown correctly</li>
            <li>Popovers now stay open when clicked and close when the user does not click inside the popover</li>
            <li>Fixed a bug where the ClinVar annotation was erroneous which happened when there were missing submissions</li>
            <li>Fixed a bug where the consequence annotation was erroneous which happened when the variant was large and spanned multiple exons / introns</li>
            <li>Fixed a bug where the user was unable to download variants which are not annotated</li>
            <li>Fixed a bug where the user could not search for class 3+</li>
            <li>The date is now inserted correctly when classifying variants. Preexisting classifications did not change (this is fixed starting from 24.10.23).</li>
        </ul>
        Known issues:
        <ul>
            <li>COSMIC links do not work sometimes</li>
        </ul>
    </div>

    <div class="underline">v 1.3 (19.06.2023)</div>
    <div>
        General changes:
        <ul>
            <li>Added checkboxes to variant table which allows more fine grained collection of variants in lists</li>
            <li>Added information about the chosen criteria to the comment of a ClinVar submission from HerediVar</li>
            <li>Updated documentation</li>
            <li>The strength of all ACMG criteria can be adjusted now</li>
            <li>Added continuous numbering of variants in variant lists</li>
            <li>Added 3+ and 3- classes. They can be selected manually from the dropdown</li>
            <li>Added important variant information to classify page</li>
            <li>Mane select transcripts are now shown in green in IGV</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Fixed ClinVar uploads <div class="classification_5_col">{{macros.draw_exclamation_mark('') }} Classifications are currently submitted to the ClinVar TEST ENDPOINT which is reset periodically! None of your ClinVar submissions will go live at the moment.</div></li>
            <li>Fixed a bug when the user tried to access a non existing variant list</li>
            <li>Improved performance of variant table</li>
            <li>Fixed a bug where the user was unable to search for gene:hgvs strings in the variant table</li>
            <li>Classified variants in IGV browser are now shown with the color scheme corresponding to the classification</li>
            <li>Fixed a bug where some characters were replaced by the url encoding in HGVS strings</li>
            <li>Fixed a couple of CSP errors</li>
            <li>Fixed the colors of classified variants in IGV</li>
            <li>Fixed colors of acmg criteria buttons</li>
        </ul>
        Known issues:
        <ul>
            <li>COSMIC links do not work sometimes</li>
        </ul>
    </div>
    <div class="underline">v 1.2 (12.05.2023)</div>
    <div>
        General changes:
        <ul>
            <li>Improved user registration form</li>
            <li>Added profile page which allows each user to edit their name, mail and affiliation by themselves</li>
            <li>Added possibility to search for multiple lists simultaneously</li>
            <li>Added duplicate, intersect, subtract and union actions for user variant lists</li>
            <li>Updated documentation</li>
        </ul>
        Bugfixes:
        <ul>
            <li>Fixed a bug where the user was able to upload multi allelic variants</li>
            <li>Fixed a bug where the user was able to insert invalid variant sequences</li>
            <li>Navigation is now shown independent of scroll position on documentation page</li>
            <li>Fixed a bug when searching it would not show any variants</li>
        </ul>
    </div>
    <div class="underline">v 1.1 (08.05.2023)</div>
    <div>
        General changes:
        <ul>
            <li>External links now open a new tab on left click</li>
            <li>Added documentation</li>
        </ul>
        Performance changes:
        <ul>
            <li>Added client-side pagination of the literature table</li>
        </ul>
    </div>
    <div class="underline">v 1.0</div>
    <div>
        Initial release
    </div>
</div>

</div>

{% endblock %}

