{% extends 'base.html' %}


{% block content %}

<div class="container">

  
<h1> {% block title %} Welcome to the HerediVar Wiki {% endblock %} </h1>




<div class="row p-relative">
    <div class="col-3 p-fixed width_medium" id="nav_container">
      <nav id="navbar-example3" class="h-100 flex-column align-items-stretch pe-4 border-end ">
        <nav class="nav nav-pills flex-column">
          <a class="nav-link" href="#item-0">General Thoughts</a>
          <a class="nav-link" href="#item-1">The workflow</a>
          <nav class="nav nav-pills flex-column">
            <a class="nav-link ms-3 my-1" href="#item-1-1">Insert variants</a>
            <a class="nav-link ms-3 my-1" href="#item-1-2">Browse variants</a>
            <a class="nav-link ms-3 my-1" href="#item-1-3">Variant details page</a>
            <a class="nav-link ms-3 my-1" href="#item-1-4">Classify a variant</a>
            <a class="nav-link ms-3 my-1" href="#item-1-5">Automatic classification algorithm: HerediClassify</a>
            <a class="nav-link ms-3 my-1" href="#item-1-6">Publish variants</a>
          </nav>
          <a class="nav-link" href="#item-2">User variant lists</a>
          <a class="nav-link" href="#item-svs">Structural variants</a>
          <a class="nav-link" href="#item-download">Download data</a>
          <a class="nav-link" href="#item-3">User management</a>
          <nav class="nav nav-pills flex-column">
            <a class="nav-link ms-3 my-1" href="#item-3-1">Create a new user</a>
            <a class="nav-link ms-3 my-1" href="#item-3-2">Change roles & disable users</a>
            <a class="nav-link ms-3 my-1" href="#item-3-3">Change password</a>
          </nav>
        </nav>
      </nav>
    </div>
  
    <div class="col-8 vbsl">
      <div data-bs-spy="scroll" data-bs-target="#navbar-example3" data-bs-smooth-scroll="true" class="scrollspy-example-2" tabindex="0">

        <div id="item-0">
          <h4>General Thoughts & Conventions</h4>
          <p>
            There are four types of users in HerediVar:
            <ol>
                <li>Superuser</li>
                <li>Standard user</li>
                <li>Read only user</li>
                <li>Anonymous user</li>
                <li>Account manager</li>
            </ol>
            An anonymous user can not view or edit any variant information on HerediVar. Anonymous users must ask an account manager to create an account for them. 
            As the name implies, the read only user can only view information. He can neither insert variants nor classify them. 
            The standard user is allwed to view and edit information on HerediVar and the superuser has access to the admin dashboard
            which allows him to import variants from HerediCare, reannotate all variants at once and do multiple other administrative tasks. <br>
            The account manager is a special role which allows the user to create and modify user data. <br>
            <br>
            Conventions for links: It is very easy to differentiate external links from internal ones: Whenever a link is displayed like {{ macros.external_link('this', '#') }} 
            it references an external website and if a link is shown like <a href="#">this</a> it references an internal page. <br>
            <br>
            It is very important to understand that each user is able to insert one classification for each variant - classification scheme combination. In addition,
            a superuser can insert one consensus classification for each variant. Consensus classifications are always superseded by the newest consensus classification.
            But all previous consensus classifications are kept and can still be viewed. 
          </p>
        </div>


        <div id="item-1">
          <h4>The workflow</h4>
          <p></p>
        </div>
        <div id="item-1-1">
          <h5>Insert variants</h5>
          <p>
            There are two ways to insert small variants. The most common one is to import variants from HerediCare. However, only superusers can do that. Standard users can 
            insert variants through the <a href="{{ url_for('variant.create')}}">create variant page</a> for small variants. New variants are automatically added to the annotation queue
            and processed once a worker is ready to process it. The status of the annotation can be seen when navigating to the variant details page. <br>
            <br>
            HerediVar also supports structural variants, however, in a limited manner. Thus, it is currently only possible to insert deletions, duplications and inversions.
            These must be inserted manually through the <a href="{{ url_for('variant.create_sv') }}"> create structural variant page</a>. Although they generally behave identical compared
            to a small variant, they are currently not annotated with meta information.
          </p>
        </div>
        <div id="item-1-2">
          <h5>Browse variants</h5>
          <p>
            To find the variant you are looking for navigate to the <a href="{{url_for('variant.search')}}">search variants page</a>. This page shows a paginated 
            table of all variants. The columns of this table are interpreted like this:
            <ol>
              <li>CHR: chromosome</li>
              <li>POS: GRCH38, left-aligned genomic position</li>
              <li>REF: reference base(s)</li>
              <li>ALT: alternative base(s)</li>
              <li>VARTYPE: the type of the variant. This value is "sv:sv_type" if the variant is a structural variant or "small" if the variant is a regular small variant</li>
              <li>HGVS c.: the HGVS c. string for this variant. The reference transcript is the "best" transcript* for the gene of this variant.</li>
              <li>HGVS p.: the HGVS p. HGVS string for this variant. The reference transcript is the "best" transcript* for the gene of this variant.</li>
              <li>GENE: the gene which has the variant. Sometimes there are multiple genes, because the variant has consequences in all of them.</li>
              <li>Your classification: the most recent classification of the user which is currently logged in.</li>
              <li>Consensus classificaiton: the most recent consensus class from the VUS-task-force.</li>
            </ol>
            *: "best" transcript refers to the preferred transcript which is calculated for each gene individually: MANE select/plus clinical -> longest coding -> longest non-coding <br> <br>
            
            To filter the table of all variants click the magnifiying glass to the top right. You can filter the table of variants by:
            <ul>
              <li>
                A variant: <br>
                The variant strings are of this format: chrom-pos-ref-alt in case of small variants and chrom-start-end-sv_type in case of structural variants
              </li>
              <li>
                The variant type: either small variants or structural variants. Checking both is the same as leaving both empty
              </li>
              <li>A gene name or unique alias or HGVS-ID (e. g. BARD1)</li>
              <li>
                A genomic range is described by providing chromosome start position and end position. These values must be separated with ":", "-" or whitespace (except new lines). Both, adding as well as removing the "chr" prefix
                from the chromosome is allowed and should yield the same result. Example formats that work:
                <ul>
                  <li>chr1:start-end</li>
                  <li>chr1-start-end</li>
                  <li>chr1 start  end</li>
                  <li>chr1[tab]start[tab]end</li>
                  <li>1:start:end</li>
                </ul>
                Examples that do not work:
                <ul>
                  <li>chr1- start-end (note the space between the first dash and the start position)</li>
                  <li>chr1::start-end</li>
                </ul>
                HerediVar allows to upload a bed file in addition to providing chromosomal ranges manually through the textarea. Simply click the "upload bed..." button and choose a bed file from your local storage.
                Keep in mind that start positions within bed files are zero based (ie. inclusive) while end coordinates are one based (ie. exclusive) within bed files. HerediVar automatically converts the endposition
                to a zero based (ie. inclusive) position. Choosing this option will automatically append all relevant fields in a HerediVar supported format to the textarea. <br>
                WARNING: Due to the fact that .bed files can contain any character each bed style range must reside in its own line (ie. they must be separated with newline characters)!
              </li>
              <li>cDNA ranges <br>
                This option allows you to filter the variants by ranges of cDNA positions. There are three required values for each query. <br>
                <ol>
                  <li>The transcript or a gene symbol. When specifiying a gene symbol the algorithm will choose the best transcript for this gene.</li>
                  <li>The cDNA start position relative to 1</li>
                  <li>The cDNA end position relative to 1</li>
                </ol>

                These values must be spearated either with a space or tabulator. <br>
                You can also scan intronic regions by adding or subtracting to the start or end cDNA position. (eg. ENST00000380152 1 68+10). Keep in mind that cDNA positions 
                are counted from the larger genomic position to the smaller one if the gene is defined on the reverse or minus strand. Still, the smaller cDNA position must be inserted as start and the bigger cDNA position as end. <br>
                If you want to scan regions outside the gene region add an asterisk (*) or a minus (-) character before the start or end position (eg. ENST00000380152 -100 1).
              </li>
              <li>A HGVS c. string <br>
                This search option is most useful when you know the reference transcript. In this case please specify the search string like this: 
                Transcript:c.PositionReference>Alternative (e. g. ENST00000260947:c.1972C>T). <br>
                If you do not know the transcript you can also specify the gene symbol (or alias or HGNC-ID) instead of the transcript (e. g. BARD1:c.1972C>T)<br>
                In case you neither know the transcript nor the gene it is also possible to omit the reference sequence. But be careful with this type of search as it 
                will provide you with ALL possible nucleotide changes that COULD be the one you are looking for. Thus, you should inspect the results carefully before continueing. (e. g. c.1972C>T)<br>
                WARNING: HerediVar will choose the ensembl mane select transcript as reference, if you specify the gene instead of the transcript. If there is no mane 
                select transcript it will omit this HGNC.
              </li>
              <li>IDs<br>
                You can search for HerediVar variant_ids or dbSNP rsids, clinvar variation ids, COSMIC ids (COSV) or heredicare vids. <br>
                You can insert the ids directly or choose to append a ":" followed by the database name. For example: You can search for rs1057519871, 1057519871:rsid 
                or 1057519871. All of which return the same variant. <br>
                It is recommended to specify the database of the identifier to make results more accurate. <br>
                Extensions used by heredivar: <br>
                <ul>
                  <li>HerediVar variant_ids: heredivar</li>
                  <li>dbSNP rsids: rsid</li>
                  <li>ClinVar variation ids: clinvar</li>
                  <li>COSMIC COSV ids: cosmic</li>
                  <li>HerediCaRe VIDs: heredicare</li>
                </ul>
              </li>
              <li>Annotations <br>
                First select the annotation from the dropdown menu, then type the operator. For numerical values you may choose from &lt;, &lt;=, &gt;, &gt;=, = and !=. 
                For text-like values you may choose from =, != and ~. Hereby, ~ is a case insensitive substring match. Finally, provide the value in the next textbox. <br>
                It is always possible to add more annotation searches. But be aware that resulting variants must always satisfy all of the criteria at the same time to appear in the variant list.
              </li>
              <li>
                A consensus classification. This filters only for consensus classifications which were made within HerediVar and excludes consensus classification which were imported from HerediCaRe.
                However, if the "include_heredicare" checkbox is ticked, the search will also display imported consensus classifications (see section on consensus classifications in documentation for further
                details on how to spot imported consensensus classifications). <br>
                Be mindful though: HerediVar native consensus classifications always overshadow imported ones. So a variant which has an imported consensus classification of 1, but a HerediVar native consensus
                classification of 2 will not be found if class 1 and include_heredicare are checked (provided class 2 is not checked).
              </li>
              <li>Your classification</li>
              <li>A list of variants</li>
            </ul>
            If multiple constraints are selected they are joined with 'AND'. This means that if you choose to search for BARD1 in the genes and select consensus classification 1 
            you will get all variants from BARD1 which are classified with 1. <br>
            You can separate multiple genes, HGVS or chromosomal ranges in their respective fields either with semicolons (';'), commas (',') or newlines. It is also possible to have a combination of them. <br>
            It is also possible to add all variants which satisfy the current filtering criteria to a user defined list by clicking the "Add to list" button located at 
            the bottom left. To create a new list go <a href="{{url_for('user.my_lists')}}">here</a>. More information about user variant lists is <a href="#item-2">here</a>.<br>
            Clicking on any of the rows will get you to its variant details page. To view details about a particular gene just click its name. 
          </p>
        </div>
        <div id="item-1-3">
          <h5>The variant details page</h5>
          <p>
            This page displays all annotations of a variant. The top headline shows the variant and its most recent consensus classification by the VUS task-force. There are multiple types of
            consensus classifications:
            <div class="d-flex">
              {% for classification in ['-', '1', '2', '3-', '3', '3+', '4M', '4', '5'] %}
              <div class="ssr">{{macros.render_consensus_classification_label(classification, 'heredivar') }}</div>
              {% endfor %}
            </div>
            These are interpreted as follows:
            <ul>
              <li>-: no classification</li>
              <li>1: benign</li>
              <li>2: likely benign</li>
              <li>3-: uncertain with tendency to benign</li>
              <li>3: uncertain</li>
              <li>3+: uncertain with tendency to pathogenic</li>
              <li>4: likely pathogenic</li>
              <li>4M: modifier</li>
              <li>5: pathogenic</li>
            </ul>
            If the consensus classification symbol shows a red download icon like so:
            <div class="d-flex">
              <div class="ssr">{{macros.render_consensus_classification_label('2', 'heredicare') }}</div>
            </div>
            It means that the consensus classification was imported from HerediCare and there is no consensus classification available for this variant which was made on HerediVar. <br>
            <div class="bsb"></div>
            Right below the headline one can find status information about the variant. The first pill shows the status of the annotation. It could be queued, processing, done or erroneous.
            Next to the annotation status pill there is the clinvar submission status. This shows if the variant was already uploaded to ClinVar from HerediVar. There is one additional, optional 
            status pill which sais "HIDDEN" if the variant is hidden. When you click the "gear" button
            the website presents options on what to do with the variant. <br>
            The main body of the page displays visualizations, classifications and annotations which are organized in multiple tabs. The tabs contain the following information:
            <ul>
              <li>Classification: HerediVar consensus classification, HerediVar user classifications, HerediCare center classifications, ClinVar classifications, BRCA exchange and ARUP BRCA classifications</li>
              <li>Population & in-silico annotations: phylop-100way, cadd scaled, revel, spliceai, maxentscan, gnomAD, flossies, cancerhotspots, tp53db, hexplorer, maxhbond score</li>
              <li>Consequence: The variant consequences with associated HGVS strings and protein domains</li>
              <li>Assays: Functional, splicing, etc. assays. Add a new assay by clicking the gear button at the top.</li>
              <li>Literature: Papers annotated by VEP, TP53db and litvar2</li>
              <li>
                IGV: The IGV genome browser. The first track shows the variant itself. The second and third track show all classified variants (click on them to find a link to the variant details page). The fourth track
                shows the ensembl transcripts. Red transcripts are MANE select or MANE plus clinical transcripts, orange ones are ensembl canonical and dark blue ones are gencode basic transcripts. Other transcripts
                which do not satisfy the aforementioned criteria are not shown.
              </li>
            </ul>
            Remember that annotations are not shown if the variant is unknown to the respective database. 
          </p>
        </div>
        <div id="item-1-4">
          <h5>Classify a variant</h5>
          <p>
            To classify a variant click the gear button and select "Classify". On the classify variant page you can first select if you want to insert a user classification or a consensus classification.
            However, only superusers can insert consensus classifications. Next, select the classification scheme or no scheme if you just want to insert a freetext classification not following a scheme.
            The webpage automatically chooses the best scheme for the variant if possible. <br>
            Click a criteria button to see more details about it. Here is an example for PVS1:
            <div>
              <img src="/static/img/documentation/classify_01.PNG" alt="missing img"  width="100%">
            </div>
            First, insert evidence and then click the selectbox next to the headline of the criterion to change its state. The following criterion states are possible:
            <ol>
                <li>unchecked: we do not have enough evidence to say anything about this criterion -> not assessed / unknown evidence</li>
                <li>selected: evidence suggests this criterion is true -> assessed / true evidence</li>
                <li>unselected: evidence suggests this criterion should not be selected -> we KNOW that this criterion is not selected because of evidence -> use this for documentation purposes</li>
            </ol>
            It is also possible to adjust the strength of some criteria. The classification preview to the right will update automatically based on the selected criteria. <br>
            Once all criteria have been assessed you can insert further evidence by adding literature. Simply add a blank row or add one from the text mining results and copy the section
            of interest. <br>
            Finally, select the final classification and optionally leave a comment. This is especially encouraged when the selected final classification differs from the proposed classification from the 
            classification scheme. <br>
            Note that each user can only have one user classification per variant and scheme, but they can be edited anytime. <br>
            Consensus classifications work just alike. The only difference is that the number of users who chose that criterion is displayed next to it like so:
            <div>
              <img src="/static/img/documentation/classify_02.PNG" alt="missing img"  width="100%">
            </div>
            The selected criterion, "PS1" was chosen by two HerediVar users to be important for this variant. The actual user input is shown in the "User selections" table. When you click the copy symbol
            it is possible to quickly copy the selection and evidence of this user. The criterion will be in an intermediate state (shown with diagonal strips). Criteria in the intermediate state simply 
            tell the user that the criterion was selected by some automatic procedure. Every intermediate criterion will be submitted just like non-intermediate criteria. So there is no need to get rid of the
            intermediate state. Simply be extra critical about intermediate criteria.
          </p>
        </div>

        <div id="item-1-5">
          <h5>Automatic classification algorithm: HerediClassify</h5>
          <p>
            The automatic classification algorithm called HerediClassify takes the evidence collected by HerediVar and automatically assesses ACMG criteria. By default it uses the ACMG SVI adaptation classification scheme,
            but it also supports gene specific ACMG classification schemes. These include:
            <ul>
              <li>ATM</li>
              <li>BRCA1</li>
              <li>BRCA2</li>
              <li>CDH1</li>
              <li>PALB2</li>
              <li>PTEN</li>
              <li>TP53</li>
            </ul>
            The best scheme is automatically selected by HerediClassify. It always calculates two classifications for the same scheme. One based on splicing evidence and the other on protein evidence.<br>
            HerediClassify is run by the annotation service of HerediVar. So a reannotation also updates the automatic classification. <br>
            The variant classify page allows to copy all selected and unselected criteria from HerediClassify along with their strength and evidence to help speed up the classification process. Using the gear button
            the user can choose if the criteria using the splicing or protein evidence should be copied. <br>
            Every criterion selected in this way will be in the intermediate state. The intermediate state is indicated with dark stripes
            across the criterion button and is meant as a visual indication for the user that the criterion might need further attention. All intermediate criteria will be treated as if they are not intermediate. This
            means that they are submitted like they are not intermediate.
          </p>
        </div>

        <div id="item-2">
          <h4>User variant lists</h4>
          <p>
            Each registered user can add new variant lists <a href="{{url_for('user.my_lists')}}">here</a>. Click the "+" button:
            <div>
              <img src="/static/img/documentation/variant_list_01.PNG" alt="missing img"  width="100%">
            </div>
            Enter a name for the list and select if it is readable / editable by all HerediVar users. If a list is readable everyone can see the list, but they can not add or remove variants from it.
            This is however possible when the list is also editable. <br>
            There are two ways to insert variants to the newly created variant list. The first one is to add variants to a list on a one-by-one basis is to navigate to the variant details page, 
            click the gear button and select "Add to list". <br>
            The second one is to visit the <a href="{{url_for('variant.search')}}">browse variants</a> page. You can simply select the variants that you wish to save to a list and select "Add selected to list".
            To simplify things one can also choose to select and unselect all variants. Keep in mind that only variants which satisfy the current search parameters are added when select all variants is active.
            The page also keeps track of selected variants through pages of the variant table. So it is save to first select all variants and then click the "Add selected to list" button once. Keep in mind though
            that your selection will reset upon querying a new search. <br>
            <br>
            All variants collected in these ways are displayed when clicked on the name of the list at the <a href="{{url_for('user.my_lists')}}">user lists page</a>. 
            The table of variants behaves exactly the same as the one shown on the <a href="{{url_for('variant.search')}}">browse variants</a> page. It can be filtered and when a row is clicked
            it takes you to the variant details page.<br>
            It is also possible to export all variants in one of the lists to vcf by clicking the "Export to VCF" button on the bottom of the page. <br>
            If you want to change the properties or delete a list that was created by you. Just click the little pencil button next to the list name. Beware that private lists can not be seen
            by other HerediVar users. <br>
            <br>
            To further simplify handling these lists there are four possible actions to manipulate variant lists. First click the list you wish to edit. Then, you can see four icon-buttons on the top
            which are explained as follows:
            <ul>
              <li>{{ macros.draw_duplicate_symbol() }}: duplicate the list</li>
              <li>{{ macros.draw_intersect_symbol() }}: intersect this list with another list</li>
              <li>{{ macros.draw_subtract_symbol() }}: subtract another list from this list</li>
              <li>{{ macros.draw_add_symbol() }}: combine this list with another list</li>
            </ul>
            As you can see, these implement simple set operations. The website will ask some more questions about how to do each of the operations. For example, some operations allow you to do it either inplace or or not in place.
            When you choose inplace it will take the result of the operation and write it to the currently selected list. When this is not selected it will create a new list (asking you to 
            choose a new name and access rights).
          </p>
        </div>

        <div id="item-1-6">
          <h4>Publish variants</h4>
          <p>
            A superuser can upload or publish variants and their annotations or classifications to external databases. Currently, HerediVar offers to upload to ClinVar and HerediCaRe.
            To publish one or multiple variants the user has to navigate to one of the following pages:
            <ul>
              <li>The variant details page: for single variant uploads</li>
              <li>The user variant lists page: uploads all variants in that user defined variant list</li>
              <li>The admin dashboard: automatically select all variants which got an update to their consensus classification since the last upload</li>
            </ul>
            The publish variants page presents a list of all variants which are about to be submitted. Simply check the database(s) below that list and fill out the required information. After the form is submitted
            your request will be processed in the background. You can get status information about each submission from the respective variant details page. <br>
            There is an automatic upload algorithm running in the background which will regularily (at least once a month) check for updated consensus classifications and upload them to HerediCaRe. <br>
            <br>
            Caveat: <br>
              It is possible to submit the same data multiple times. Though, HerediVar will skip the upload request and show a red triangle within the respective status pill on the variant details page.
              Consider this case: A user made a fresh consensus classification and submitted that to HerediCaRe. Right after he notices that the consensus classification needs adjustments. So he 
              creates another consensus classification and submits that as well. In this case the updated consensus classification WILL NOT BE SUBMITTED, but skipped and the red triangle will show up. <br>
              This however will be resolved once a user resubmits the consensus classification after the previous one is finished or once the automatic upload algorithm catches the updated consensus classification.
              So whenever this happens consensus classifications take more time to fully update in the third party database.
          </p>
        </div>
        
        
        <div id="item-svs">
          <h4>Structural variants</h4>
          <p>
            Structural variants can only be inserted using the <a href="{{ url_for('variant.create_sv') }}">create structural variants page</a>. They generally behave identical to a small variant, but are not annotated by
            the annotation service. Also, the interpretation of the information in the browse variants table differs slightly:
            <div>
              <img src="/static/img/documentation/structural_variants_01.PNG" alt="missing img"  width="100%">
            </div>
            The differences are:
            <ul>
              <li>If a deletion is longer than 1000 bp, the POS column contains the start and end position of the structural variant separated by a dash (-). 
                The sequence of duplications and inversions is never displayed, independent of the length of the variant.</li>
              <li>The POS column will show a red triangle if the position is imprecise</li>
              <li>The HGVS c. and p. columns contain the hgvs strings that were provided by the user upon creation of the variant. They are not ordered in any way nor are they curated or correspond to the genes shown in the gene column</li>
              <li>The GENE column shows only genes that overlap the breakpoints of the structural variants</li>
            </ul>
            The variant details page of structural variants is also similar to that of small variants with the only difference being that the variant banner at the top shows all genes that overlap the region of the structural variant.
            These are color coded:
            <ul>
              <li>blue: gene overlaps breakpoint</li>
              <li>bold: VUS-Task-Force important gene</li>
              <li>black: non breakpoint overlapping gene</li>
            </ul>
          </p>
        </div>


        <div id="item-download">
          <h4>Download data</h4>
          <p>
            Anonymous users can download all classified variants and all classified structural variants in vcf <a href="{{ url_for('main.downloads') }}">here</a>. They will only see the consensus classifications by the VUS-task-force along
            with some meta information. Most of this data is also publically available through ClinVar. <br>
            <br>
            Registered users are able to download a vcf file for one variant through the respective variant details page by clicking the gear button and selecting "Export to VCF". This file will contain all information that are available
            for this variant. Including all annotations and classifications. <br>
            <br>
            If a registered user wants to download more than one variant in a single file they have to collect relevant variants in a user defined list (see previous section on how to create one). Then, simply click the "Export list to VCF"
            button at the bottom right of the list details page.  Note that structural variants and small variants can be mixed in this case. <br>
            <br>
            In rare cases ClinVar submissions are missing although there is a ClinVar summary available. In such cases the clinvar_submissions info in the downloaded vcf file will be set to None. This is usually because of erroneous data hosted
            by ClinVar and is often fixed by updating ClinVar.<br>
            <br>
            In any case, downloaded VCF files are encoded using the url encoding (<a href="https://en.wikipedia.org/wiki/Percent-encoding">Wikipedia</a>). Excluding whitespace, which is replaced with underscores "_". This ensures that the
            file is machine readable and consistent when special characters that are also used as delimiters are used within the actual information. It is fast forward to revert the encoding once relevant information was read from the file.
            Many programming languages support url decoding out of the box.<br>
            <br>
            The downloaded VCF files from HerediVar always adhere to this delimiter order:
            <ol>
              <li>;</li>
              <li>&</li>
              <li>|</li>
              <li>$</li>
              <li>+</li>
            </ol>
          </p>
        </div>


        <div id="item-3">
          <h4>User management</h4>
          <p></p>
        </div>
        <div id="item-3-1">
          <h5>Create a new user</h5>
          <p>
            Only users with the "account_manager" role are allowed to create new users. There is no possibility to self-register in HerediVar.
            An account manager can go <a href="{{ url_for('auth.register') }}">here</a> to create a new user account. Simply fill out the form which is presented 
            to you. Keep in mind that the username must be unique and can not be changed later. The mail address must be unique as well.<br>
            Upon submission Heredivar will automatically notify the user that about the newly created account using the provided e-mail. The new user will 
            recieve two mails: one with the username specified and one with an automatically generated initial password.
          </p>
        </div>
        <div id="item-3-2">
          <h5>Change roles & disable users</h5>
          <p>
            An account manager can edit existing users by visiting <a href="{{url_for('auth.show_users')}}">{{url_for('auth.show_users')}}</a>. Simply select a user from the list
            to edit its information. It is possible to disable users or change their roles from this page. <br>
            Keep in mind that changeing the e-mail will require the user to verify the mail upon next login.
          </p>
        </div>
        <div id="item-3-2">
          <h5>Change password</h5>
          <p>
            Every user can change their own password by visiting <a href="{{url_for('auth.change_password')}}">this</a> page. This will redirect to the keycloak account management
            console. First click the Update-button:
            <div>
              <img src="/static/img/documentation/change_password.PNG" alt="missing img"  width="100%">
            </div>
            Then, you are probably asked to re-authenticate using your old password. After that you can insert a new password.
          </p>
        </div>


      </div>
    </div>
  </div>
  





{% endblock %}

{% block special_scripts %}
<script src="/static/js/documentation.js"></script>
{% endblock %}